use imagemagick and libraw for raw image support

uhthomas opened the pull request:
imagemagick and libraw have generally good support for raw images, including
Sony's ARW format. These tools should also allow Immich to support many more
image formats in future without any major code changes.

https://www.libraw.org/supported-cameras

I've tested and verified this change with .ARW files and other standard formats.

Fixes: https://github.com/immich-app/immich/issues/2156

uhthomas commented:
I've tested this, and am awaiting a response from the sharp author regarding https://github.com/lovell/sharp/issues/3612#issuecomment-1576714726. Please don't merge yet.

samip5 commented:
I think the idea was to use darktable.

See https://github.com/immich-app/immich/pull/2198

uhthomas commented:
> I think the idea was to use darktable.
> 
> See #2198

Also see https://github.com/immich-app/immich/issues/2156#issuecomment-1561955606.

uhthomas commented:
It would be nice to merge https://github.com/immich-app/immich/pull/2736 before this.

uhthomas commented on server/Dockerfile:
```diff
@@ -17,13 +17,13 @@ RUN npm run build
 WORKDIR /usr/src/app
 
-RUN apk add --no-cache vips-heif vips vips-cpp ffmpeg perl
+RUN apk add --no-cache ffmpeg imagemagick-dev libraw-dev perl vips-dev vips-heif vips-magick
```
`-dev` packages may not be needed here.

uhthomas replied:
I tested without `vips-dev` and it wasn't happy at all. I presume all it wants is `vips-cpp`.

uhthomas replied:
Nope, not happy with `vips-cpp` either. I'll leave the `-dev` packages as is :)

uhthomas commented on server/src/domain/media/media.service.ts:
```diff
@@ -60,9 +60,9 @@ export class MediaService {
+      } catch (error: any) {
         this.logger.warn(
-          `Failed to generate jpeg thumbnail using sharp, trying with exiftool-vendored (asset=${asset.id})`,
+          `Failed to generate jpeg thumbnail using sharp, trying with exiftool-vendored (asset=${asset.id}): ${error.message}`,
```
Q for reviewers: Should we just remove this? vips/imagemagick should do this automatically. Might be nice to reduce code footprint and complexity. What do you think?

jrasm91 replied:
This was a work around (fallback) for raw images that had an embedded thumbnail. If raw photos work without this I think we can definitely remove it.

uhthomas replied:
https://github.com/immich-app/immich/issues/2744

jrasm91 reviewed (commented):
How big of an image size jump does moving away from alpine cause?

uhthomas commented:
> How big of an image size jump does moving away from alpine cause?

Hm? We are using alpine.

uhthomas commented on web/src/lib/utils/asset-utils.ts:
```diff
@@ -131,28 +131,20 @@ export function getFileMimeType(file: File): string {
+			'.nef': 'image/x-nikon-nef',
+			'.raf': 'image/x-fuji-raf',
+			'.srw': 'image/x-samsung-srw'
+		}[getFilenameExtension(file.name)] ?? ''
```
You're right, I'm not sure how I did that. Thank you.

uhthomas replied:
@jrasm91 I did one even better and added a test for this. The test also revealed that null or undefined values were not handled correctly, and has been fixed.

jrasm91 commented:
> > How big of an image size jump does moving away from alpine cause?
> 
> Hm? We are using alpine.

Ah, sorry I mis-read the change. `alpine` got bumped to the next line with the addition of the hash.

elliotclee reviewed (commented):
Hi,

Thanks for working on this. I like the way you changed to using arrays instead of endless if() statements in a number of areas.

I think it would make sense to add support for all the filetypes that libraw can handle. I came looking for .cr2 (image/x-canon-cr2), and I gather there are several other types that it would be helpful to accept. See https://invent.kde.org/graphics/digikam/-/blob/master/core/libs/rawengine/drawfiles.cpp for a fairly complete list of the file types supported by libraw, as well as https://github.com/angryziber/gnome-raw-thumbnailer/blob/master/data/raw-thumbnailer.xml for info on the MIME types for them.

Hope this helps,

uhthomas commented:
> Hi,
> 
> Thanks for working on this. I like the way you changed to using arrays instead of endless if() statements in a number of areas.
> 
> I think it would make sense to add support for all the filetypes that libraw can handle. I came looking for .cr2 (image/x-canon-cr2), and I gather there are several other types that it would be helpful to accept. See https://invent.kde.org/graphics/digikam/-/blob/master/core/libs/rawengine/drawfiles.cpp for a fairly complete list of the file types supported by libraw, as well as https://github.com/angryziber/gnome-raw-thumbnailer/blob/master/data/raw-thumbnailer.xml for info on the MIME types for them.
> 
> Hope this helps,

I would love to do this, but I am afraid for the scope of this PR becoming too large, which is why I separated some changes into #2736. I think the best thing to do would be to merge this as-is, closing the original Sony ARW issue and create another issue like #2743 to support the remaining libraw formats. Does that sound good to you? I can create that issue now.

jrasm91 commented:
I think it is OK to add, test, and verify support for specific formats. With the change to the base image it'll be pretty trivial to add support for additional formats. Whether we do all that testing in this PR or it follow-up PRs seems less important. I don't think @uhthomas needs to commit to testing every single format type before we merge this support for ARW.

uhthomas commented:
I've created #2745 to track the implementation of all libraw formats.

uhthomas commented:
It looks like this job was cancelled and is confusing GitHub. Any idea what happened, or if it's intentional?

https://github.com/immich-app/immich/actions/runs/5248322532/jobs/9479643740?pr=2668

uhthomas commented:
Strange, I made force pushed and now a different build was 'cancelled' https://github.com/immich-app/immich/actions/runs/5248561828/jobs/9480220088?pr=2668.

jrasm91 commented:
I have had some issues of late with force-pushing having weird impact on the running jobs.

uhthomas commented:
Ah, https://www.githubstatus.com/incidents/csnbwc85chp2.

uhthomas committed:
use imagemagick and libraw for raw image support

imagemagick and libraw have generally good support for raw images, including
Sony's ARW format. These tools should also allow Immich to support many more
image formats in future without any major code changes.

https://www.libraw.org/supported-cameras

I've tested and verified this change with .ARW files and other standard formats.

Fixes: #2156

alextran1502 committed:
Add additional type for awr

uhthomas commented on web/src/lib/utils/asset-utils.ts:
```diff
@@ -125,34 +125,20 @@ export function getAssetFilename(asset: AssetResponseDto): string {
-	}
+	const mimeTypes: Record<string, string> = {
+		'3gp': 'video/3gpp',
+		arw: 'image/x-sony-arw' || 'image/awr',
```
I don't think that will work?

alextran1502 committed:
pr feedback

uhthomas commented:
It seems like this image is missing quite a few runtime dependencies. Namely imagemagick-dev and vips-magick.

https://github.com/immich-app/immich/blob/af86712df3b093cc2a342cff3279833ba055d520/server/Dockerfile#L26
